curvefs/src/space/bitmap_allocator.cpp

```diff
@@ -0,0 +1,379 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * Project: curve
+ * File Created: Fri Jul 16 21:22:40 CST 2021
+ * Author: wuhanqing
+ */
+
+#include "curvefs/src/space/bitmap_allocator.h"
+
+#include <glog/logging.h>
+
+#include <algorithm>
+#include <vector>
+
+#include "curvefs/src/space/fast_align.h"
+
+namespace curvefs {
+namespace space {
+
+namespace {
+
+inline bool HasAllocOffsetHint(const SpaceAllocateHint& hint) {
+    return hint.leftOffset != hint.INVALID_OFFSET ||
+           hint.rightOffset != hint.INVALID_OFFSET;
+}
+
+inline bool IsAllocTypeBig(const SpaceAllocateHint& hint) {
+    return hint.allocType == AllocateType::BIG;
+}
+
+}  // namespace
+
+uint64_t BitmapAllocator::CalcBitmapAreaLength(
+    const BitmapAllocatorOption& opt) {
+    uint64_t len = align_down<uint64_t>(
+        opt.length * (1 - opt.smallAllocProportion), opt.sizePerBit);
+
+    return std::min(len, opt.length);
+}
+
+struct BitmapAllocator::AllocOrder {
```

Comment:
why not define in bitmap_allocator.h?

Reply:
AllocOrder is private in BitmapAllocator, so it only for internal usage, no need to define it in header file.